tiled/adapters/hdf5.py

```diff
@@ -114,7 +115,10 @@ def from_uri(
         access_policy=None,
     ):
         filepath = path_from_uri(data_uri)
-        file = h5py.File(filepath, "r", swmr=swmr, libver=libver)
+        cache_key = (cls.__module__, cls.__name__, filepath)
```

Comment:
Just a thought... I wonder about implications of keying the resources by `cls.__name__` and `cls.__module__`... Could there be situations when the same file is cached from different adapters at (roughly) the same time? (Probably not a big issue for reading). But this key definitely has great observability and traceability.
Another possible option could be to delegate the creation of cache keys to the `with_resource_cache` function itself (possibly with an option to override), so that whoever writes the adapter doesn't need to worry about keys at all. Then a key could be a hash of `filename`, all `*args`, and all `*kwargs` (but again, it might create problems if multiple threads would be trying to access a file concurrently).

Reply:
My concern was two adapters stepping on each other, maybe even accessing the same file in some weird scenario but caching different types of objects, say from different I/O libraries like tifffile versus imageio. If Adapter A caches a type of object and then Adapter B gets it from the cache this could lead to very confusing bugs.

But maybe I am overengineering this for a very weird hypothetical case. I like your idea of making the cache key optional, and by default hashing the inputs.

Reply:
Potential collisions with other `kwargs` is a really good point; I haven't thought about it tbh. Though, I agree, it's probably rather unlikely, especially if we call it `_cache_key` with underscore (or even `_resource_cache_key` to be extra safe, but probably unnecessary).
I personally would be in favour of option 2, as it adds the least overhead on the caller of `with_resource_cache`; option 3 is nice, but I feel it could be too complicated. Actually, if there are any clashes with `kwargs` in option 2, one can always resort to `functools.partial`.

Reply:
On one hand, I like the idea of implementing the `cache_key` in the framework, taking control of criticl piece of infrastructure that is so very easy to get wrong and has terrible consequences when it's wrong.

On the other hand,  I also like the idea of the adapter developer having to produce a `cache_key` that is appropriate for their adapter's underlying file type. I have already seen stupid use cases for adapters a am little skeptical of our ability to produce the right `cache_key` for all cases.

Reply:
Another option: mash up these ideas by keeping the explicit `cache_key` but generally using the pattern:

```py
cache_key = (f, args, sorted(kwargs.items())
x = with_resource_cache(cache_key, f, *args, **kwargs)
```

rather than my `(type(self).__module__, type(self).__name__, ...)` pattern.

Reply:
Despite the risk of abuse / poor choices, I think leaving the cache key generation up to the adapter is the right choice.  Trying to do something generic by typesniffing + ensure that kwargs are hashable (see some of the wild performance issues we found it intake) is also something that is very easy to get wrong and "sometimes framework does it, sometimes the handlers do it" makes thing harder to debug / track down.

If we expect 100s-1000s of adapters to exist / to be provided by the end-users scientists then maybe I see the case for the framework doing it, but we see <100 adapters mostly written by controls / computing group staff I think making sure the cache key is "good" is easier and clearer in the long run.

Reply:
Decision then, with @genematx on Zoom: Adapter must specify the key. We'll update the cache key to be more specific so that potentially different adapters can share cache items if they are both opening the same file with an identical function call.